### PR TITLE
Add support for port aggregation

### LINKED ISSUE #142: Add switch port aggregation support
We have a US-24-POE switch that uses link aggregation for a couple of devices. It'd be very nice to be able to configure that setup as part of Terraform, however presently that doesn't appear to work.

I hit "apply" in a similar configuration to the screenshot below, this is the example request that was set to the UDM-P that I use as my controller:

```
URI: https://unifi.my.network/proxy/network/api/s/default/rest/device/5fa75def5318bd035aa209bd
METHOD: PUT
{
    "port_overrides": [
        {
            "port_idx": 11,
            "portconf_id": "5f3d11ec5318bd051a842c6e",
            "port_security_mac_address": [],
            "op_mode": "aggregate",
            "aggregate_num_ports": 2
        },
        {
            "port_idx": 9,
            "portconf_id": "5f3d11ec5318bd051a842c6e",
            "op_mode": "aggregate",
            "port_security_mac_address": [],
            "aggregate_num_ports": 2
        }
    ]
}
```

On my machine the portconf_id there is the default "All" configuration as seen here from my terraform state:

```
data "unifi_port_profile" "all" {
    id   = "5f3d11ec5318bd051a842c6e"
    name = "All"
    site = "default"
}
```

And the device in the request URI is the US-24-POE, also as seen from my terraform state:

```
# unifi_device.us_24_poe:
resource "unifi_device" "us_24_poe" {
    disabled = false
    id       = "5fa75def5318bd035aa209bd"
    mac      = "de:vi:ce:ma:cn:um"
    site     = "default"

    port_override {
        number          = 11
        port_profile_id = "5f3d11ec5318bd051a842c6e"
    }
    port_override {
        number          = 9
        port_profile_id = "5f3d11ec5318bd051a842c6e"
    }
}
```

This doesn't seem too difficult to add, alas I don't know golang very well so I'm not of much use here :)

Let me know if you need any additional details.

Screenshot of relevant configuration:

image

### PR DESCRIPTION
**Pull request by paultyng**
Fixes #142

### CODE REVIEW

**Comment by paultyng** [comment]
Unsure how to test this, need to see if its possible without adopting a device or if it needs one to run against.

File: internal/provider/resource_device.go
Comment (joshuaspence) [inline internal/provider/resource_device.go:92]: Should we add validation to make sure this is only set if `op_mode` is set to `aggregate`?

**Review by joshuaspence — APPROVED** [review]
It doesn't need to be a part of this PR, but I think we should add acceptance test coverage that is opt-in with an environment variable to run against real hardware. Something like `UNIFI_SWITCH=<MAC_ADDRESS>`.

**Comment by dekarl** [comment]
Is there an easy way to test this in combination with gitlab-terraform images?

I built the plugin and simply replaced the cached one in .terraform/providers/registry.terraform.io/paultyng/unifi/0.34.0/linux_amd64/terraform-provider-unifi_v0.34.0 but on the next run of gitlab-terraform plan it gets replaced with a pristine copy from the registry...

The change itself looks good when I compare it to the JSON returned when configuring a LAG.

Maybe add some validation, like aggregate_num_ports must be set to use op_mode=aggregate. Or that there is no port_override for port 12 if there is a 2-port-LAG on port 11.

**Comment by dekarl** [comment]
I managed to test it locally, appears to be working.
A small nit, the port override to have just one native VLAN on switching port 3 detects a drift with nothing to apply every time I run terraform plan. Looks like aggregate_num_ports should be omitted when op_mode=switching or aggregate_num_ports = 0

```
Terraform detected the following changes made outside of Terraform since the last "terraform apply":

  # unifi_device.switch-sued-2 has been changed
  ~ resource "unifi_device" "switch-sued-2" {
        id       = "5fde1f005ab20c014e3f7d19"
        name     = "sued-switch-2"
        # (3 unchanged attributes hidden)

      + port_override {
          + aggregate_num_ports = 0
          + name                = "switch-maas"
          + number              = 3
          + port_profile_id     = "61a80d3d5ab20c0144a62015"
        }
      - port_override {
          - name            = "switch-maas" -> null
          - number          = 3 -> null
          - port_profile_id = "61a80d3d5ab20c0144a62015" -> null
        }
        # (1 unchanged block hidden)
    }

```

**Comment by dekarl** [comment]
I'm using this branch successfully in homelab-production. But I'd love to return to the released version. What is still needed before merging it?

**Comment by CarbonCollins** [comment]
:wave: Is planned to be merged at any point? Would very much like to be able to configure this :D

**Comment by szinn** [comment]
@paultyng Anything else blocking this from being merged?

**Comment by joshuaspence** [comment]
@dekarl @szinn @CarbonCollins I've added test coverage and merged this now. Please feel free to test it out and report any issues you find. I will test it myself and make a new release in the next few days

**Pull request merged**
